packetbeat/protos/cassandra/cassandra.go

```diff
@@ -0,0 +1,200 @@
+package cassandra
+
+import (
+	"time"
+
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/logp"
+
+	"github.com/elastic/beats/packetbeat/protos"
+	"github.com/elastic/beats/packetbeat/protos/tcp"
+	"github.com/elastic/beats/packetbeat/publish"
+)
+
+// cassandra application level protocol analyzer plugin
+type cassandra struct {
+	ports        protos.PortsConfig
+	parserConfig parserConfig
+	transConfig  transactionConfig
+	pub          transPub
+}
+
+// Application Layer tcp stream data to be stored on tcp connection context.
+type connection struct {
+	streams [2]*stream
+	trans   transactions
+}
+
+// Uni-directioal tcp stream state for parsing messages.
+type stream struct {
+	parser parser
+}
+
+var (
+	debugf = logp.MakeDebug("cassandra")
+
+	// use isDebug/isDetailed to guard debugf/detailedf to minimize allocations
+	// (garbage collection) when debug log is disabled.
+	isDebug = false
+)
+
+func init() {
+	protos.Register("cassandra", New)
+}
+
+// New create and initializes a new cassandra protocol analyzer instance.
+func New(
+	testMode bool,
+	results publish.Transactions,
+	cfg *common.Config,
+) (protos.Plugin, error) {
+	p := &cassandra{}
+	config := defaultConfig
+	if !testMode {
+		if err := cfg.Unpack(&config); err != nil {
+			return nil, err
+		}
+	}
+
+	if err := p.init(results, &config); err != nil {
+		return nil, err
+	}
+	return p, nil
+}
+
+func (cassandra *cassandra) init(results publish.Transactions, config *cassandraConfig) error {
+	if err := cassandra.setFromConfig(config); err != nil {
+		return err
+	}
+	cassandra.pub.results = results
+
+	isDebug = logp.IsDebug("http")
+	return nil
+}
+
+func (cassandra *cassandra) setFromConfig(config *cassandraConfig) error {
+
+	// set module configuration
+	if err := cassandra.ports.Set(config.Ports); err != nil {
+		return err
+	}
+
+	// set parser configuration
+	parser := &cassandra.parserConfig
+	parser.maxBytes = tcp.TCP_MAX_DATA_IN_STREAM
+
+	// set parser's compressor, only `snappy` supported right now
+	if config.Compressor == Snappy {
+		parser.compressor = SnappyCompressor{}
+	} else {
+		parser.compressor = nil
+	}
```

Comment:
right, we will need some new buffer to put compressed information into. But I think the parser as is supports decompression (due to managing an internal read buffer). 

In case you want to handle decompression yourself: If compression can not span multiple frames (need to check protocol this is true), it's quite straight forward to use `p.buf.Collect()` to collect the compressed frames into `[]byte` or better wrap `p.buf` with `LimitReader` and pass the reader to `zlib.Reader` (the reader allocated from zlib is resettable with a new reader, such that no new buffers need to be allocated -> reduces garbage to be collected). If using the later approach check for possibly unread bytes in presence of potential padding, though.

From CQL doc:

> As of this version 2 of the protocol, the following compressions are available:
    - lz4 (https://code.google.com/p/lz4/). In that, note that the 4 first bytes
      of the body will be the uncompressed length (followed by the compressed
      bytes).
    - snappy (https://code.google.com/p/snappy/). This compression might not be
      available as it depends on a native lib (server-side) that might not be
      avaivable on some installation.

That is, if lz4 is used, we do know the required buffer size for decompressed content before decompression. Same goes for snappy:

> The stream starts with the uncompressed length (up to a maximum of 2^32 - 1),
stored as a little-endian varint. Varints consist of a series of bytes,
where the lower 7 bits are data and the upper bit is set iff there are
more bytes to be read. In other words, an uncompressed length of 64 would
be stored as 0x40, and an uncompressed length of 2097150 (0x1FFFFE)
would be stored as 0xFE 0xFF 0x7F.

Will check tomorrow how/if decompression can be handled when doing code review.